// @vivliostyle/cli/dist/chunk-PBAUN5CU.js
import {
  Logger,
  useTmpDirectory
} from "./chunk-MDTA37GZ.js";

// src/config/vite.ts
import upath from "upath";
import {
  createLogger,
  mergeConfig as mergeViteConfig,
  resolveConfig
} from "vite";
import { dim } from "yoctocolors";
var defaultConfigEnv = {
  preview: { command: "serve", mode: "development", isPreview: false },
  build: { command: "serve", mode: "production", isPreview: true }
};
async function resolveViteConfig({
  serverRootDir,
  server,
  viteConfig,
  viteConfigFile,
  logLevel,
  workspaceDir,
  mode
}) {
  const viteLogger = createLogger(
    {
      silent: "silent",
      info: "info",
      verbose: "info",
      debug: "info"
    }[logLevel],
    { allowClearScreen: false }
  );
  const warnedMessages = /* @__PURE__ */ new Set();
  viteLogger.info = (msg) => {
    Logger.logInfo(`${dim("[vite]")} ${msg}`);
  };
  viteLogger.warn = (msg) => {
    viteLogger.hasWarned = true;
    Logger.logWarn(`${dim("[vite]")} ${msg}`);
  };
  viteLogger.warnOnce = (msg) => {
    if (warnedMessages.has(msg)) {
      return;
    }
    viteLogger.hasWarned = true;
    Logger.logWarn(`${dim("[vite]")} ${msg}`);
    warnedMessages.add(msg);
  };
  viteLogger.error = (msg) => {
    viteLogger.hasWarned = true;
    Logger.logError(`${dim("[vite]")} ${msg}`);
  };
  const root = typeof serverRootDir === "string" ? serverRootDir : (await useTmpDirectory())[0];
  const finalUserConfig = mergeViteConfig(viteConfig || {}, {
    server,
    preview: server,
    configFile: viteConfigFile === true ? void 0 : viteConfigFile,
    root,
    customLogger: viteLogger,
    cacheDir: upath.join(workspaceDir, ".vite")
  });
  return await resolveConfig(
    finalUserConfig,
    defaultConfigEnv[mode].command,
    defaultConfigEnv[mode].mode,
    "development",
    defaultConfigEnv[mode].isPreview
  );
}

export {
  resolveViteConfig
};
