// @vivliostyle/cli/dist/chunk-4IIM6RSG.js
// src/const.ts
import fs from "node:fs";
import { fileURLToPath } from "node:url";
import resolvePkg from "resolve-pkg";
import upath from "upath";
var MANIFEST_FILENAME = "publication.json";
var TOC_FILENAME = "index.html";
var TOC_TITLE = "Table of Contents";
var COVER_HTML_FILENAME = "cover.html";
var COVER_HTML_IMAGE_ALT = "Cover image";
var XML_DECLARATION = '<?xml version="1.0" encoding="UTF-8"?>';
var EPUB_OUTPUT_VERSION = "3.0";
var EPUB_NS = "http://www.idpf.org/2007/ops";
var EPUB_CONTAINER_XML = `${XML_DECLARATION}
<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
  <rootfiles>
    <rootfile full-path="EPUB/content.opf" media-type="application/oebps-package+xml" />
  </rootfiles>
</container>`;
var EPUB_LANDMARKS_TITLE = "Landmarks";
var EPUB_LANDMARKS_TOC_ENTRY = TOC_TITLE;
var EPUB_LANDMARKS_COVER_ENTRY = "Cover Page";
var EMPTY_DATA_URI = "data:,";
var VIEWER_ROOT_PATH = "/__vivliostyle-viewer";
var CONTAINER_URL = "ghcr.io/vivliostyle/cli";
var CONTAINER_ROOT_DIR = "/data";
var CONTAINER_LOCAL_HOSTNAME = "host.docker.internal";
var cliRoot = upath.join(fileURLToPath(import.meta.url), "../..");
var cliVersion = (() => {
  if (import.meta.env?.VITEST) {
    return "0.0.1";
  }
  const pkg = JSON.parse(
    fs.readFileSync(upath.join(cliRoot, "package.json"), "utf8")
  );
  return pkg.version;
})();
var viewerRoot = resolvePkg("@vivliostyle/viewer", { cwd: cliRoot });
var coreVersion = (() => {
  if (import.meta.env?.VITEST) {
    return "0.0.1";
  }
  const pkg = JSON.parse(
    fs.readFileSync(upath.join(viewerRoot, "package.json"), "utf8")
  );
  return pkg.version;
})();

export {
  MANIFEST_FILENAME,
  TOC_FILENAME,
  TOC_TITLE,
  COVER_HTML_FILENAME,
  COVER_HTML_IMAGE_ALT,
  XML_DECLARATION,
  EPUB_OUTPUT_VERSION,
  EPUB_NS,
  EPUB_CONTAINER_XML,
  EPUB_LANDMARKS_TITLE,
  EPUB_LANDMARKS_TOC_ENTRY,
  EPUB_LANDMARKS_COVER_ENTRY,
  EMPTY_DATA_URI,
  VIEWER_ROOT_PATH,
  CONTAINER_URL,
  CONTAINER_ROOT_DIR,
  CONTAINER_LOCAL_HOSTNAME,
  cliVersion,
  viewerRoot,
  coreVersion
};
